doc/guides/modules.rst

```diff
@@ -44,6 +44,9 @@ Module Repositories
 * The module repository codebase shall include a *module.yml* file in a
   :file:`zephyr/` folder at the root of the repository.
 
+* The *module.yml* file shall contain the ``name`` key and have the value
```

Comment:
This is only necessary if the module's name is different from the project name by default, right? So `shall` feels a bit too strong, unless we write something like this:

```
For inclusion in upstream Zephyr, the module's name shall match the ``name``
field for the project in :zephyr_file:`west.yml`.
```

Reply:
no, because per default the `path` name is used for module name.

The origin of this, is because `west` is optional, and the module must behave consistently regardless of whether `west` is used for obtaining the modules using `west` or if the modules are passed to the build system using: `cmake -DZEPHYR_MODULES=<module> ...` or `cmake -DZEPHYR_EXTRA_MODULES=<module> ...`

If the `west` project names were used, then the module would behave differently when passed on command line or when fetched using `west`.

Reply:
Therefore, I propose `shall` to ensure consistent behavior with `west`, also when using a module without `west` as some users do.

Reply:
Regarding:
`For inclusion in upstream Zephyr, `
then the section this text is placed in is actually covering the modules in Zephyr.
https://docs.zephyrproject.org/latest/guides/modules.html#module-repositories

and as I already commented here: https://github.com/zephyrproject-rtos/zephyr/pull/27185#discussion_r469309273, it rightfully belongs in the contribution page.

Reply:
> The origin of this, is because `west` is optional, and the module must behave consistently regardless of whether `west` is used for obtaining the modules using `west` or if the modules are passed to the build system using: `cmake -DZEPHYR_MODULES=<module> ...` or `cmake -DZEPHYR_EXTRA_MODULES=<module> ...`

You can still extract the name from the path if west is not being used. This has never been a problem.

This IMO will cause some major breakge with existing modules if we start requiring the name to be specified for the modules to work with west.

Reply:
> You can still extract the name from the path if west is not being used. This has never been a problem.
> 
> This IMO will cause some major breakge with existing modules if we start requiring the name to be specified for the modules to work with west.

Code is fully backward compatible, so if no name is specified in the in `zephyr/module.yml` then path is used.

The name from `west` will never be used, only the name in `zephyr/module.yml`, so if module `foo` has `zephyr/module.yml` with content:
```
name: foo
```
then that module will always have the name `foo`, both when using `ZEPHYR_MODULES` and when using `west`, regardless of the project name in `west` (which theoretical could be `bar`).

Thus the reason for this comment: https://github.com/zephyrproject-rtos/zephyr/pull/31156#issuecomment-756371408 to specify those two names should be in sync.

Reply:
> Code is fully backward compatible, so if no name is specified in the in `zephyr/module.yml` then path is used.

Right, which is why I think we just need the `name` in module.yml if the path is different from the west.yml name, and `shall` is too strong, right?

Reply:
> 
> 
> > Code is fully backward compatible, so if no name is specified in the in `zephyr/module.yml` then path is used.
> 
> Right, which is why I think we just need the `name` in module.yml if the path is different from the west.yml name, and `shall` is too strong, right?

Of course there might be different opinions here, but let me give some reasons why I think shall should be used for Zephyr maintained modules:
1. Using same principle for all modules makes things consistent (all modules have the `name` key)
2. Combined with #30904 it will be clear how to refer to a modules name, as example, with today's solution:
   west project name: `hal_nordic`
   github repo: https://github.com/zephyrproject-rtos/hal_nordic
   path (when using `west`): `modules/hal/nordic`
   Zephyr module name: `nordic`, meaning `ZEPHYR_NORDIC_MODULE_DIR` is the variable identifying the module.
   Combining this with #30904 results in the folder `ZEPHYR_BASE/modules/nordic` for glue code.
   If we enforce all modules to have a name key, such inconsistencies doesn't appear.
3. With #30904 it becomes more important how modules are named, as there is now a need to determine a modules name. Before #30904, it generally didn't matter what name a module had, but that is changing.
4. It allows to freely restructure the project layout, and not having to consider how this might impact glue code, or any other build code.
5. When not using `west`, someone might decide to place, as example `modules/hal/nordic` directly in a `hal_nordic` folder, causing the modules name to change, and this could impact glue code in Zephyr repo.
   When glue code was located in the module itself, it almost didn't matter what the module was, as it was rarely used (currently only `ZEPHYR_TFM_MODULE_DIR` and `ZEPHYR_MCUBOOT_MODULE_DIR` was used in upstream Zephyr), but with glue code, and the need to identify if a module is present when including the glue code, a known module name is much more important.
6. Will easier allow us to make `name` key mandatory in later release, as all Zephyr modules will be up-to-date.
   Starting with a shall, allows us to slowly update modules, when making other changes anyway.

As an additional note, we also have a *shall* here:
> The module repository codebase shall include a module.yml file in a zephyr/ folder at the root of the repository.

even though a `zephyr/module.yml` is not mandatory for inclusion of a module.
It is sufficient to have `zephyr/CMakeLists.txt` or `zephyr/Kconfig`. Having one of those two files present at this folder also identifies a Zephyr module.

Reply:
The consistency really doesn't seem worth losing the single point of truth when one is available to me.